[dip-196] Initial onchain network identity DIP draft

gregnazario opened the pull request:
### Summary

DIP for Onchain Network Identity

gregnazario commented:
@aching @davidiw Could I get some feedback on this DIP?  I followed DIP 0 as the instructions, but the process has changed.  I can update the instructions if we need to

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+# Simple Summary
+
+This DIP describes Onchain Network, an identity for Diem nodes to easily discover and update identity of other Diem nodes on the network.  Allows for trusted connections between nodes.
```
Hmm.... This sentence seems a bit lacking:

This DIP describes Onchain Network Identities, a framework for Diem nodes to easily discover and update network credentials and permissions for other nodes on the network. Examples of such permissions include priority or privileged access for well-known participants, such as VASPs, to upstream validator fullnodes. This becomes increasingly important as a means to mitigate resource limitations of the validator fullnodes, where the validator fullnodes cannot support the demands of the entire ecosystem but can still support nodes from entities critical for network operations. While in theory this can be handled without the blockchain leveraging an identity repository and configuration file manipulation, the requirements for scaling and securing such a solution would be substantially more operational overhead and duplicate functionality already provided by the Diem blockchain.

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+import useBaseUrl from '@docusaurus/useBaseUrl';
+
+# Simple Summary
```
Probably no need to have summary, abstract, and motivation, maybe just one section?

gregnazario replied:
We should update DIP-0 appropriately.

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+# Abstract
+
+Currently, network discovery is done either manually (via configuration) or by the validator set (which only applies to validators and their respective full nodes).  Adding onchain network identity allows for nodes to discover other nodes via the shared onchain resources, and allow owners of the nodes to control their identity and updates to the identity. This DIP describes a method of storing a  ...
+
+
+# Motivation
+
+This method will be used first for the Association to connect their full nodes to the validator full node network, with uninterrupted access.  Other nodes may choose to use this method for discovery in the future.
```
delete

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+A consumer may choose to subscribe to all accounts, in which it will watch the `Treasury compliance` account updates and keep track of which accounts to watch accordingly.
+
+# Rationale
```
delete?

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+The motivation of the `NetworkIdentity` design is to keep identity updates to be controlled by the owner of the account.  Additionally, this allows for a more distributed `NetworkIdentity` that can be used for other purposes.  An onchain list of accounts was not used, as either an offchain account list can be used, or a consumer can listen to the `Treasury compliance` account updates.
+
+# Backwards Compatibility
```
delete

davidiw replied:
why keep :P

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+As this only applies to discovery of future nodes, there are no concerns for backwards compatibility.
+
+# Implementations
```
implementation should never be in this DIP, delete this from dip0

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+# Implementations
+
+N/A
+
```
recommend an Examples section of how the Diem tech stack would leverage this -- e.g., event emitting, pick up identity, update networking stack... mention other enhancements that you built out to help mitigate attacks (TBF and connection limits)

gregnazario replied:
This isn't specified in DIP-0.  We should update the process even if it's just suggested.

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,70 @@
+
+The `NetworkIdentity` for an account shall be updated by only the owner of the account.  At this point, the consumers of this identity must update their local use of identity within a reasonable amount of time.
+
+## Consuming Network Identity
+
+Consumers of `NetworkIdentity` may keep track of a list of accounts to watch.  The consumers then can watch the event streams of each account's `NetworkIdentity` for updates and update accordingly.
+
+A consumer may choose to subscribe to all accounts, in which it will watch the `Treasury compliance` account updates and keep track of which accounts to watch accordingly.
```
can you provide the code snippet?

gregnazario replied:
Currently reworking some of the technical details involving integration with move.  Can come up with examples and code snippets once that's done.

davidiw reviewed (commented):
Coming along well

gregnazario commented:
@aching Do you have any other comments for this DIP?

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+the associated network identity so that the nodes can trust them.  This DIP describes the ability for node owners to
+self register their network identity onchain, and allow other nodes to lookup this data for discovery of other nodes.
+An example of a use case for this is partners guaranteeing access to Validator Full Nodes.  Validator full nodes can
+easily subscribe and know which nodes to prioritize.
```
```suggestion
subscribe to an event stream to learn about changes and determine which nodes to prioritize.
```

davidiw replied:
avoid superlatives

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+
+# Summary
+
+Diem nodes need the ability to find other nodes in the network easily.  Additionally, the nodes need the ability to find 
```
```suggestion
Dip-6 defines a method for downstream nodes to learn about validator full nodes and gain access to the Diem network.  In turn, validator full nodes would benefit by being able to distinguish known nodes from unknown.
```

davidiw replied:
I don't really buy the value immediately from being able to discover other nodes other than VFNs.

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+
+## Consuming Network Identity
+
+Consumers of `NetworkIdentity` may keep track of a list of accounts to watch.  The consumers then can watch the event streams of each account's `NetworkIdentity` for updates and update accordingly.
```
please update and show the actual event type

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+# Backwards Compatibility
+
+As this only applies to discovery of future nodes, there are no concerns for backwards compatibility.
+
```
Can we add a blurb about applications that tie back into your earlier TBF and connection limitation work?

gregnazario replied:
As a way to say how this works together with that work?  Or as a way to justify why VFNs need this identity?

aching commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+
+Diem nodes need the ability to find other nodes in the network easily.  Additionally, the nodes need the ability to find 
+the associated network identity so that the nodes can trust them.  This DIP describes the ability for node owners to
+self register their network identity onchain, and allow other nodes to lookup this data for discovery of other nodes.
```
```suggestion
self register their network identity on-chain, and allow other nodes to lookup this data for discovery of other nodes.
```

aching replied:
Per David's comment above, the on-chain network identify seems to only be a feature for VFNs.  It might make this DIP more straightforward to focus on that aspect unless there are other concrete use cases you would like to call out.

gregnazario replied:
Sure, I can focus on only the VFN usecase

aching commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+# Summary
+
+Diem nodes need the ability to find other nodes in the network easily.  Additionally, the nodes need the ability to find 
+the associated network identity so that the nodes can trust them.  This DIP describes the ability for node owners to
```
Is node owner a formal term?  E.g. in DIP-6 Validator Owner?  If so we should still to the terminology or if generic, perhaps node operator might be a more appropriate.

gregnazario replied:
Yeah, can change to `node operator`, the differences between `owner` and `operator` are blurred then at the full node usage.

aching reviewed (commented):
Nice work!

davidiw commented on dips/dip-196.md:
```diff
@@ -0,0 +1,55 @@
+
+As this only applies to discovery of future nodes, there are no concerns for backwards compatibility.
+
+# Copyright Notice
+
+This documentation is made available under the Creative Commons Attribution 4.0 International (CC BY 4.0) license (available at https://creativecommons.org/licenses/by/4.0/).
```
actually I would argue the other dips screwed up :D

gregnazario replied:
* Under [DIP requirements in DIP-0](https://github.com/diem/dip/blob/main/dips/dip-0.md#dip-requirements) the copyright notice is required.  It also says, it's at the bottom and well it's not :/
* This is present in a few DIPs, see [DIP-1](https://github.com/diem/dip/blob/main/dips/dip-1.mdx#copyright-notice).
* The DIP website adds the copyright notice to the bottom of every DIP, regardless of whether it has it at the bottom.  See [DIP-160 website](https://dip.diem.com/dip-160/) vs [DIP-160 git](https://github.com/diem/dip/blob/main/dips/dip-160.md)

Let's please determine what we need to fix in DIP-0 and correct this process to remove the confusion.

aching replied:
Sounds good.  Let's just keep it consistent with DIP-1 and we'll work on how to revise DIP-0 or outdate it.

aching commented:
Have you had a chance to get to the feedback?  I think once addressed, we can commit the draft and make more progress here.

gregnazario commented:
> Have you had a chance to get to the feedback? I think once addressed, we can commit the draft and make more progress here.

I've addressed all the feedback but examples / code snippets, as I'm still waiting on changes in the move language to support vec<vec<u8>>
